aptos-move/framework/aptos-stdlib/sources/pool_u64.move

```diff
@@ -0,0 +1,188 @@
+    public fun create(shareholders_limit: u64): Pool {
+        Pool {
+            shareholders_limit,
+            total_coins: 0,
+            total_shares: 0,
+            shares: simple_map::create<address, u128>(),
+            shareholders: vector::empty<address>(),
+        }
+    }
+
+    /// Destroy an empty pool. This will fail if the pool has any balance of coins.
+    public fun destroy_empty(pool: Pool) {
+        assert!(pool.total_coins == 0, error::invalid_state(EPOOL_IS_NOT_EMPTY));
+        let Pool {
+            shareholders_limit: _,
+            total_coins: _,
+            total_shares: _,
+            shares: _,
+            shareholders: _,
+        } = pool;
+    }
+
+    /// Return `pool`'s total balance of coins.
+    public fun total_coins(pool: &Pool): u64 {
+        (pool.total_coins as u64)
+    }
+
+    /// Return the total number of shares across all shareholders in `pool`.
+    public fun total_shares(pool: &Pool): u64 {
+        (pool.total_shares as u64)
+    }
+
+    /// Return true if `shareholder` is in `pool`.
+    public fun contains(pool: &Pool, shareholder: address): bool {
+        simple_map::contains_key(&pool.shares, &shareholder)
+    }
+
+    /// Return the number of shares of `stakeholder` in `pool`.
+    public fun shares(pool: &Pool, shareholder: address): u64 {
+        if (contains(pool, shareholder)) {
+            (*simple_map::borrow(&pool.shares, &shareholder) as u64)
+        } else {
+            0
+        }
+    }
+
+    /// Return the balance in coins of `shareholder` in `pool.`
+    public fun balance(pool: &Pool, shareholder: address): u64 {
+        let num_shares = shares(pool, shareholder);
+        shares_to_amount(pool, num_shares)
+    }
+
+    /// Return the list of shareholders in `pool`.
+    public fun shareholders(pool: &Pool): vector<address> {
+        pool.shareholders
+    }
+
+    /// Return the number of shareholders in `pool`.
+    public fun shareholders_count(pool: &Pool): u64 {
+        vector::length(&pool.shareholders)
+    }
+
+    /// Update `pool`'s total balance of coins.
+    public fun update_total_coins(pool: &mut Pool, new_total_coins: u64) {
+        pool.total_coins = (new_total_coins as u128);
+    }
+
+    /// Allow an existing or new shareholder to add their coins to the pool in exchange for new shares.
+    public fun buy_in(pool: &mut Pool, shareholder: address, coins_amount: u64): u64 {
+        let num_shares = amount_to_shares(pool, coins_amount);
+        pool.total_coins = pool.total_coins + (coins_amount as u128);
+        pool.total_shares = pool.total_shares + (num_shares as u128);
+        add_shares(pool, shareholder, num_shares);
+
+        num_shares
+    }
+
+    /// Add the number of shares directly for `shareholder` in `pool`.
+    /// This would dilute other shareholders if the pool's balance of coins didn't change.
+    public fun add_shares(pool: &mut Pool, shareholder: address, num_shares: u64): u64 {
```

Comment:
this shouldn't be public, imo

Reply:
Why not? This allows diluting existing shareholders, which is a very commonly used feature. The caller of this needs to have to control of their Pool (mutable ref required) so they're definitely the owner.

Reply:
like below, I don't understand what this does, diluting for what reason? maybe we can start private and make it public?

this goes to the lack of a description of what pool is trying to provide.

Reply:
There's a description added in https://github.com/aptos-labs/aptos-core/pull/3836